src/ui/api/projectmember.go

```diff
@@ -0,0 +1,216 @@
+	pma.Data["json"] = make([]models.Member, 0)
+	if pma.id == 0 {
+		//member id not set, return all member of current project
+		memberList, err := project.GetProjectMember(queryMember)
+		if err != nil {
+			log.Errorf("Failed to query database for member list, error: %v", err)
+			pma.RenderError(http.StatusInternalServerError, "Internal Server Error")
+			return
+		}
+		if len(memberList) > 0 {
+			pma.Data["json"] = memberList
+		}
+	} else {
+		//return a specific member
+		queryMember.ID = pma.id
+		memberList, err := project.GetProjectMember(queryMember)
+		if err != nil {
+			log.Errorf("Failed to query database for member list, error: %v", err)
+			pma.RenderError(http.StatusInternalServerError, "Internal Server Error")
+			return
+		}
+		if len(memberList) > 0 {
+			pma.Data["json"] = memberList[0]
+		}
+	}
+	pma.ServeJSON()
+}
+
+// Post ... Add a project member
+func (pma *ProjectMemberAPI) Post() {
+	projectID := pma.project.ProjectID
+	var request models.MemberReq
+	pma.DecodeJSONReq(&request)
+	err := AddOrUpdateProjectMember(projectID, request)
+	if err != nil {
+		log.Errorf("Failed to add project member, error: %v", err)
+		pma.RenderError(http.StatusInternalServerError, "Failed to add project member")
+		return
+	}
+}
+
+// Put ... Update an exist project member
+func (pma *ProjectMemberAPI) Put() {
+	pid := pma.project.ProjectID
+	pmID := pma.id
+	if pmID == 0 {
+		log.Errorf("Failed to update DB to add project user role, project id: %d, pmid : %d", pid, pmID)
+		pma.RenderError(http.StatusInternalServerError, "Failed to update data in DB")
+		return
+	}
+	var req models.Member
+	pma.DecodeJSONReq(&req)
+	err := project.UpdateProjectMemberRole(pmID, req.Role)
+	if err != nil {
+		log.Errorf("Failed to update DB to add project user role, project id: %d, pmid : %d, role id: %d", pid, pmID, req.Role)
+		pma.RenderError(http.StatusInternalServerError, "Failed to update data in DB")
+		return
+	}
+}
+
+// Delete ...
+func (pma *ProjectMemberAPI) Delete() {
+	pmid := pma.id
+	err := project.DeleteProjectMemberByID(pmid)
+	if err != nil {
+		log.Errorf("Failed to delete project roles for user, project member id: %d, error: %v", pmid, err)
+		pma.RenderError(http.StatusInternalServerError, "Failed to update data in DB")
+		return
+	}
+}
+
+// AddOrUpdateProjectMember ... If the project member relationship does not exist, create it. if exist, update it
+func AddOrUpdateProjectMember(projectID int64, request models.MemberReq) error {
+	var member models.Member
+	member.ProjectID = projectID
+	member.Role = request.Role
+	if request.MemberUser.UserID > 0 {
+		member.EntityID = request.MemberUser.UserID
+		member.EntityType = common.UserMember
+	} else if request.MemberGroup.ID > 0 {
+		member.EntityID = request.MemberGroup.ID
+		member.EntityType = common.GroupMember
+	} else if len(request.MemberUser.Username) > 0 {
+		member.EntityType = common.UserMember
+		userID, err := auth.SearchAndOnBoardUser(request.MemberUser.Username)
```

Comment:
This should be moved to API layer, package `dao` should only about accessing the DB.